### PR TITLE
Catch interest accrual and update interest overflow

### PR DESCRIPTION
**Pull request by prat-gpt**
# Description of change
## High level
* Add try catch for interest accrual and update interest method to avoid pool locking.
* Add tearDown to unit tests where overflow was occurring, so that we know pool can be recovered after such overflows.

### CODE REVIEW

**Review by mattcushman — COMMENTED** [review]
Should we similarly wrap the updateInterestState call?

File: src/base/Pool.sol [hunk 1]
```diff
@@ -561,19 +561,22 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
...
-                reserveAuction.totalInterestEarned += newInterest;
+                ) returns (uint256 newInflator, uint256 newInterest) {
+                    poolState_.inflator = newInflator;
+                    // After debt owed to lenders has accrued, calculate current debt owed by borrowers
+                    poolState_.debt = Maths.wmul(poolState_.t0Debt, poolState_.inflator);
+
+                    // update total interest earned accumulator with the newly accrued interest
+                    reserveAuction.totalInterestEarned += newInterest;
+                } catch {
+                    emit InterestOverflow();
```
Comment (grandizzy) [inline src/base/Pool.sol]: should we set `poolState_.isNewInterestAccrued` to false here too?

Comment (grandizzy) [inline src/base/Pool.sol]: I wouldn't include `overflow` in naming as there could be other reasons and also is not important the cause if failure, maybe better smth like `InterestUpdateFailure` or so?

File: tests/forge/unit/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 2]
```diff
@@ -1253,7 +1254,8 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
         skip(13 hours);
 
         // Revert with Arithmetic overflow in `Maths.wmul(pendingFactor - Maths.WAD, poolState_.debt)` in accrue interest
```
Comment (brianmcmichael) [inline tests/forge/unit/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol]: Comment here probably needs an update if it's not reverting.

File: tests/forge/unit/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 3]
```diff
@@ -1109,7 +1109,7 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
         });
     }
 
-    function testAccruePoolInterestRevertDueToExpLimit() external {
+    function testAccruePoolInterestRevertDueToExpLimit() external tearDown {
```
Comment (brianmcmichael) [inline tests/forge/unit/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol]: Test name should be changed if expecting an emit instead of a revert.

**Comment by brianmcmichael** [comment]
Really need to think about the side-effects of this one. If a pool is bust for a high inflator you'd want to allow repay/withdrawal, but avoid allowing further borrows against the pool, which probably should revert.

File: src/base/Pool.sol (hunk 1 above)
Comment (prat-gpt) [reply src/base/Pool.sol]: agreed. will update.

**Comment by prat-gpt** [comment]
> Should we similarly wrap the updateInterestState call?

this is already being done [here](https://github.com/ajna-finance/contracts/pull/954/files#diff-b8a3c7e9053439915e5f4dc0ecc997ebfa0e8197de2ca64c1a83531c24333a7aR685)

**Review by EdNoepel — COMMENTED** [review]
While this is a technically elegant solution, I am concerned about impacts to the game theory.  Does this incentivize borrowers to intentionally brick the pool, such that they can continue borrowing without paying interest?  If so, how could they do so?

Should we only wrap the following calls to catch interest overflow?
- **Removal of liquidity and unencumbered collateral from buckets** - Lenders should be able to exit.
- **Repayment of debt and pulling collateral** - Borrowers should be allowed to exit.
- **Kicking new liquidations and settlement of liquidations** - Borrowers would otherwise not be incentivized to repay.

Unsure of whether we should catch on the following operations:
- **Adding liquidity/collateral to buckets** - Should arbitrageurs be able to purchase overpriced deposit?

These changes seem to warrant further discussion.

**Comment by prat-gpt** [comment]
> While this is a technically elegant solution, I am concerned about impacts to the game theory. Does this incentivize borrowers to intentionally brick the pool, such that they can continue borrowing without paying interest? If so, how could they do so?
> 
> Should we only wrap the following calls to catch interest overflow?
> 
> * **Removal of liquidity and unencumbered collateral from buckets** - Lenders should be able to exit.
> * **Repayment of debt and pulling collateral** - Borrowers should be allowed to exit.
> * **Kicking new liquidations and settlement of liquidations** - Borrowers would otherwise not be incentivized to repay.
> 
> Unsure of whether we should catch on the following operations:
> 
> * **Adding liquidity/collateral to buckets** - Should arbitrageurs be able to purchase overpriced deposit?
> 
> These changes seem to warrant further discussion.

Also should we allow further borrows against the pool or not as mentioned by @brianmcmichael

**Comment by prat-gpt** [comment]
I think we should catch all the actions, if lenders are not gaining any interest they will start removing their liquidity from the pool which will lead to kicking of borrower loans. Borrowers will know that the lenders will be moving their liquidity out of the pool so they should also exit the pool and not make any further borrows.

**Review by mattcushman — APPROVED** [review]
lgtm

**Review by EdNoepel — APPROVED** [review]
Per today's discussion, I'm happy with the non-opinionated implementation which avoids revert on all actions.  Borrowers will be disincentivized to draw debt by both the origination fee and the risk of being `lenderKick`ed by liquidity providers trying to exit the pool.

File: src/base/Pool.sol [hunk 4]
```diff
@@ -561,19 +561,23 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
...
-                reserveAuction.totalInterestEarned += newInterest;
+                ) returns (uint256 newInflator, uint256 newInterest) {
+                    poolState_.inflator = newInflator;
+                    // After debt owed to lenders has accrued, calculate current debt owed by borrowers
+                    poolState_.debt = Maths.wmul(poolState_.t0Debt, poolState_.inflator);
+
+                    // update total interest earned accumulator with the newly accrued interest
+                    reserveAuction.totalInterestEarned += newInterest;
+                } catch {
+                    poolState_.isNewInterestAccrued = false;
```
Comment (ith-harvey) [inline src/base/Pool.sol:578]: Hmm, is this line necessary? Seems that `isNewInterestAccrued == true` if enough time elapsed and a call was made on the pool again -> https://github.com/ajna-finance/contracts/pull/954/files#diff-b8a3c7e9053439915e5f4dc0ecc997ebfa0e8197de2ca64c1a83531c24333a7aR560

Let me know if I'm missing something

Comment (EdNoepel) [reply src/base/Pool.sol:578]: If either debt or the inflator overflows, we want the pool state to reflect that interest has not accrued, undoing the optimistic assignment on line 560.

~~Another way to do it would be to change the conditional on line 564 to check `elapsed != 0`, and move the assignment of `poolState_.isNewInterestAccrued` inside the `try` block,~~ but I am happy with the implementation in this PR. 

Edit: Actually, there is no `try` block in solidity, it would need to be refactored into an awkward function to assign it in that manner.

**Review by ith-harvey — APPROVED** [review]
LGTM
